samples/nrf9160/aws_fota/src/main.c

```diff
@@ -0,0 +1,571 @@
+			       (unsigned int)sizeof(struct sockaddr_in),
+			       (unsigned int)sizeof(struct sockaddr_in6));
+		}
+
+		addr = addr->ai_next;
+		break;
+	}
+
+	/* Free the address. */
+	freeaddrinfo(result);
+}
+
+static int provision(void)
+{
+	int err;
+
+	err = tls_credential_add(CONFIG_NRF_CLOUD_SEC_TAG,
+				 TLS_CREDENTIAL_CA_CERTIFICATE,
+					AmazonRootCA3_pem,
+					AmazonRootCA3_pem_len);
+	if (err < 0) {
+		printk("Failed to register ca certificate: %d",
+		       err);
+		return err;
+	}
+	err = tls_credential_add(CONFIG_NRF_CLOUD_SEC_TAG,
+				 TLS_CREDENTIAL_PRIVATE_KEY,
+				 private_pem_key,
+				 private_pem_key_len);
+	if (err < 0) {
+		printk("Failed to register private key: %d",
+		       err);
+		return err;
+	}
+	err = tls_credential_add(CONFIG_NRF_CLOUD_SEC_TAG,
+				 TLS_CREDENTIAL_SERVER_CERTIFICATE,
+				 public_pem,
+				 public_pem_len);
+	if (err < 0) {
+		printk("Failed to register public certificate: %d",
+		       err);
+		return err;
+	}
+
+	return 0;
+}
+
+static int provision_modem_key(void)
+{
+	static sec_tag_t sec_tag_list[] = {CONFIG_NRF_CLOUD_SEC_TAG};
+
+#if defined(CONFIG_BSD_LIBRARY)
+	{
+		int err;
+
+		/* Delete certificates */
+		nrf_sec_tag_t sec_tag = CONFIG_NRF_CLOUD_SEC_TAG;
+
+		for (nrf_key_mgnt_cred_type_t type = 0; type < 3; type++) {
+			err = nrf_inbuilt_key_delete(sec_tag, type);
+			printk("nrf_inbuilt_key_delete(%lu, %d) => result=%d",
+				sec_tag, type, err);
+		}
+
+		/* Provision CA Certificate. */
+		err = nrf_inbuilt_key_write(CONFIG_NRF_CLOUD_SEC_TAG,
+					NRF_KEY_MGMT_CRED_TYPE_CA_CHAIN,
+					AmazonRootCA3_pem,
+					AmazonRootCA3_pem_len);
+		if (err) {
+			printk("NRF_CLOUD_CA_CERTIFICATE err: %d", err);
+			return err;
+		}
+
+		/* Provision Private Certificate. */
+		err = nrf_inbuilt_key_write(
+			CONFIG_NRF_CLOUD_SEC_TAG,
+			NRF_KEY_MGMT_CRED_TYPE_PRIVATE_CERT,
+			private_pem_key,
+			private_pem_key_len);
+		if (err) {
+			printk("NRF_CLOUD_CLIENT_PRIVATE_KEY err: %d", err);
+			return err;
+		}
+
+		/* Provision Public Certificate. */
+		err = nrf_inbuilt_key_write(
+			CONFIG_NRF_CLOUD_SEC_TAG,
+			NRF_KEY_MGMT_CRED_TYPE_PUBLIC_CERT,
+				 public_pem,
+				 public_pem_len);
+		if (err) {
+			printk("NRF_CLOUD_CLIENT_PUBLIC_CERTIFICATE err: %d",
+				err);
+			return err;
+		}
+	}
+#else
+  #error "Missing CONFIG_BSD_LIB"
+#endif
+}
+
+/* Function to get the client id */
+static int client_id_get(char *id)
+{
+#if !defined(NRF_CLOUD_CLIENT_ID)
+#if defined(CONFIG_BSD_LIBRARY)
+	int at_socket_fd;
+	int bytes_written;
+	int bytes_read;
+	char imei_buf[IMEI_LEN + 1];
+	int ret;
+
+	at_socket_fd = nrf_socket(NRF_AF_LTE, 0, NRF_PROTO_AT);
+	__ASSERT_NO_MSG(at_socket_fd >= 0);
+
+	bytes_written = nrf_write(at_socket_fd, "AT+CGSN", 7);
+	__ASSERT_NO_MSG(bytes_written == 7);
+
+	bytes_read = nrf_read(at_socket_fd, imei_buf, IMEI_LEN);
+	__ASSERT_NO_MSG(bytes_read == IMEI_LEN);
+	imei_buf[IMEI_LEN] = 0;
+
+	snprintf(id, CLIENT_ID_LEN + 1, "nrf-%s", imei_buf);
```

Comment:
Where does this "nrf-<IMEI>" convention come from? Should it be in a module, maybe add a comment?

Reply:
This is ripped out of `nrf_cloud` just to get the same client_id as used by `nrf_cloud` so that we make sure that our cert+id matches.

Reply:
Will leave it for now but add a comment about it.